### PR TITLE
New URL Phishing Pack

### PR DESCRIPTION
**Pull request by yguezpa**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
Script that run the URL phishing model

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 5.5.0
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:546]: I'm not sure we need this as result always, maybe add "debug" argument and only when it set to True return it

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:548]: same here

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:34]: It's our default name for the phishing model so it's not good.
Please change it to "url_phishing_model"
Also, set it as automation argument with a default

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:21]: Just to make sure I understand - everytime we will update the model on the docker, you will also update this variable in the script?

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:17]: please use non capital letters - usually all capital letters is for consts so it's bit confusing

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:459]: remove commented code

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:470]: it's better to check the entry type with the function `is_error` and to return the error if it's apply
something like: `if is_error(res): return_error(get_error(res))`

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:499]: this is the wrong tag - change to DBOT_URL_PHISHING_..

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:485]: Url -> URL
also change all of the below
you can also export this to a variable

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:483]: what happends if we have few URLs? 
I think we need to return the result only for 1 of them (maybe make it as argument) which is the most likely to be phishing.
Mabye also few results is fine - let's ask Olga

File: Packs/Base/Scripts/DBotUpdateLogoURLPhishing/DBotUpdateLogoURLPhishing.py
Comment (erezh31) [inline Packs/Base/Scripts/DBotUpdateLogoURLPhishing/DBotUpdateLogoURLPhishing.py:1]: I think this script share too much functionality with the other one.
Maybe we should you the original one with special arguments and this script will just do "executeCommand" with those special arguments. Let's talk about it

**Review by erezh31 — CHANGES_REQUESTED** [review]

**Comment by ShirleyDenkberg** [comment]
@yguezpa In the Release Notes, please add the Cortex XSOAR version for each of the scripts.
@erezh31 @lperryp Doc review completed.

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (yguezpa) [reply Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:21]: Each time we will deliver new model/change in code we will change the major. If user update the model we will change the minor

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (yguezpa) [reply Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:470]: sometime it's not error but just a string.

File: Packs/Base/Scripts/DBotUpdateLogoURLPhishing/DBotUpdateLogoURLPhishing.py
Comment (yguezpa) [reply Packs/Base/Scripts/DBotUpdateLogoURLPhishing/DBotUpdateLogoURLPhishing.py:1]: The code seems to be similar but there are many small difference. I think it's more convenient to have 2 separate scripts rather than running one from the other one.

Comment (erezh31) [reply Packs/Base/Scripts/DBotUpdateLogoURLPhishing/DBotUpdateLogoURLPhishing.py:1]: It can be harder to maintain, because most of the code is similar and duplicated - so if you change something in the logic you have to change it on both scripts

Comment (lperryp) [reply Packs/Base/Scripts/DBotUpdateLogoURLPhishing/DBotUpdateLogoURLPhishing.py:1]: + it's on Base, it means all customers will see this script and will try to understand what it does, might be confusing as it's kind of helper for another script

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (lperryp) [inline Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:272]: Can this get into a loop? 
+ if I try to use the script with a url which doesn't exist anymore, I get an error from the script without an explanation 
for instance, 
`!DBotPredictURLPhishing urls="www.wala.co.il"`

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (lperryp) [inline Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:458]: This flow is a bit problematic, as a lot of keys are missing from context compared to when the model is forced. Please add all keys from full entry, and you can specify a `Unknown` value. also need a field for verdict which is not human readable so it'd be easy to work with, just the verdict (without formatting)
image

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (lperryp) [inline Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:307]: also here, add `URLScore` and `Final Verdict` outputs which are not HR

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (lperryp) [inline Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:228]: not sure anythong good can be done here, but just to take into account that sometimes the rasterized url is request to collect personal data 
image

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (lperryp) [inline Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:532]: @erezh31 currently the model appears in the ML models page, and it seems like dev6 is updated, we need to add another config? 
image

image

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (lperryp) [inline Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:81]: Remove space please from value so it won't cause issues

**Review by lperryp — CHANGES_REQUESTED** [review]
looks great, some comments for organizing the output

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (yguezpa) [reply Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:81]: No it's for HR not context here

Comment (lperryp) [reply Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:81]: image

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (yguezpa) [reply Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:458]: This is the first entry we return with only 2 fields. There are others entries with details which should contains same fields. This is the code:

def return_entry_white_list(url):
    """
    Create syntethci entry when url belongs to whitelist
    :param url: url
    :return:
    """
    explain = {
        KEY_CONTENT_DOMAIN: extract_domainv2(url),
        KEY_CONTENT_URL: url,
        KEY_CONTENT_AGE: MSG_WHITE_LIST,
        KEY_CONTENT_LOGO: MSG_WHITE_LIST,
        KEY_CONTENT_LOGIN: MSG_WHITE_LIST,
        KEY_CONTENT_URL_SCORE: MSG_WHITE_LIST,
        KEY_CONTENT_SEO: MSG_WHITE_LIST
    }
    explain_hr = {
        "Domain": extract_domainv2(url),
        "Has the domain good SEO ?": MSG_WHITE_LIST,
        DOMAIN_AGE_KEY: MSG_WHITE_LIST,
        "Has the website a login page?": MSG_WHITE_LIST,
        "Logo found that does not correspond to the given domain": MSG_WHITE_LIST,
        "URL severity score (from 0 to 1)": MSG_WHITE_LIST
    }
    verdict_hr = {
        "Verdict": BENIGN_VERDICT,
        "URL": url
    }
    return_entry = {
        "Type": entryTypes["note"],
        "ContentsFormat": formats['json'],
        "HumanReadable": tableToMarkdown("Verdict", verdict_hr) + tableToMarkdown("Report", explain_hr),
        "Contents": explain,
        "EntryContext": {'DBotPredictURLPhishing': explain}
    }
    demisto.results(return_entry)

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (yguezpa) [reply Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:307]: no it's 2 different entries. we can talk about it

File: Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (yguezpa) [reply Packs/Base/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:228]: unfortunately nothing ...

**Review by glicht — CHANGES_REQUESTED** [review]
Please add a README

**Comment by ShahafBenYakir** [comment]
@erezh31 Will review when the PR with the right pack will be opened.

File: Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:639]: when you check if the model exists - please also get the model data itself, so you won't need to call "getMLModel" again

File: Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:670]: here you suppose to have already model_64_str from "oob_model_exists_and_updated" function

File: Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:672]: Please do this step first, the URL extraction - so you will know if there is no URL URL to process - you don't even to load the model....

**Review by erezh31 — CHANGES_REQUESTED** [review]
Please make sure you call "getMLModel" once
Let's try also to improve the model logic itself and make it more efficient, also make the model smaller:
1. On the domain lookup - make sure you stored the whitelist as "set" and not as "list" so the lookup will be fast. We can store 30K top domains I think it should be enough
2. For logo detection - maybe let's choose 10 logos
3. Please sned me the model code so I can take a look also if we can make it more efficient

File: Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (yguezpa) [reply Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:639]: done

File: Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (yguezpa) [reply Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:670]: done

File: Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (yguezpa) [reply Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:672]: Not possible since the URL extraction will check if some URL belongs to whitelist and the whitelist is inside the model itself.

File: Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:681]: add comment - "update the entire model when not exists, or when we update the model and the customer hasn't change the logos"

File: Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:686]: add comment here "loading existing model"

File: Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:689]: add comment here - "update the model with the new version and keep the customer custom logos

File: Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (erezh31) [inline Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:704]: Maybe wrap this all logic in a function that return the "urls"

**Review by erezh31 — APPROVED** [review]

**Comment by erezh31** [comment]
@yguezpa let's also wait for approval from content team
@ShahafBenYakir

File: Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (DeanArbel) [inline Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:727]: Please add a `try` and `except` blocks here, and handle the errors being raised from the script inside the `except` block. If you're not sure exactly how to handle them, you can use other scripts / integrations in content as examples. `HelloWorld` can be a good example.

File: Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.yml
Comment (DeanArbel) [inline Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.yml:115]:
```suggestion
fromversion: 6.0.0
```

**Review by DeanArbel — CHANGES_REQUESTED** [review]
Good job, the script looks good, though there are a lot of notes regarding coding standards that we enforce in our content items.

Please make sure to fix any warnings / errors in the build. If you need assistance in doing that, feel free to contact me.

File: Packs/PhishingURL/README.md
Comment (yguezpa) [reply Packs/PhishingURL/README.md:9]: Done will add the doc after

File: Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (yguezpa) [reply Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:127]: ok will change all of them. Any particular reason ?

File: Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (yguezpa) [reply Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:227]: done

File: Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (yguezpa) [reply Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:32]: sure will use format in the future

File: Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py
Comment (yguezpa) [reply Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:689]: ok any reason for that ? @DeanArbel

Comment (DeanArbel) [reply Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:689]: Mainly for coding standards, but also because `demisto.results` cannot be used with `CommandResults`.

Comment (yguezpa) [reply Packs/PhishingURL/Scripts/DBotPredictURLPhishing/DBotPredictURLPhishing.py:689]: ok thanks

**Review by DeanArbel — APPROVED** [review]
Looks good.

**Review by abaumgarten — APPROVED** [review]
Secrets check passed locally.

**Review by tomer-pan — APPROVED** [review]

**Comment by ShirleyDenkberg** [comment]
@lperryp @erezh31 @DeanArbel @abaumgarten @tomer-pan @glicht @anara123 
Doc review completed.

**Pull request merged**
